### ISSUE TITLE
Add C API

### ISSUE THREAD

**Issue by asinghvi17** [earth-mover/icechunk#381]
It would be great to have a C-accessible API for Icechunk, as a "lowest common denominator" for access from Julia, C, GDAL, etc. 

I am personally a contributor to Julia's Zarr implementation in https://github.com/JuliaIO/Zarr.jl, and would be happy to write a Julia integration for Icechunk given a C API from this end.  Interop from Julia to Rust directly is tricky (it's a similar problem as C++ has) so a C-API is the easiest way to go here.

I imagine it could be structured pretty simply, the store would likely have to be represented as a pointer.  C doesn't look to have many async capabilities but that can be handled on the calling language's end (Julia has async calls, for example, that live in a specific pool of threads).

Would be great to hear what might be needed to support this!

**Comment by rabernat** [comment]
Yes, we think this would be a great addition and support it 💯 percent!

It would be great to find a contributor who has experience creating C bindings from Rust libraries to work on this. @asinghvi17 is that something you have any experience with?

**Comment by meggart** [comment]
I have some limited experience in creating C bindings for a Rust library I have written 5 years ago for an internal project and for calling the lib from R, python, Julia and Matlab. However as this is quite some time ago I would need to check if what I have done back then is still idiomatic. 

I think a good step to get us started would be if you could share a small hello-world-icechunk example as a standalone Rust application, maybe one that just opens an existing public dataset and requests a chunk of data. I or someone else if there are volunteers could then try to create an MWE that starts wrapping the subset of functions called in this example, creates c headers automatically and calls the functions from Julia. Then we at least get a first idea of the problems we might face and the amount of work it would be to get to complete C bindings.

**Comment by paraseba** [comment]
@meggart we have a couple of example applications in the `examples` [directory](https://github.com/earth-mover/icechunk/tree/main/icechunk-python/examples). [smoke_test.py](https://github.com/earth-mover/icechunk/blob/main/icechunk-python/examples/smoke-test.py) is a pretty trivial one in there. That would give you an understanding of the basic python bindings.

If you want something Rust only, we also have an `examples` dir. In there I'd recommend [this one](https://github.com/earth-mover/icechunk/blob/main/icechunk/examples/multithreaded_store.rs).

**Comment by mpiannucci** [comment]
I am happy to provide some guidance here as well. The simplest implmentation of this could start with porting the zarr store interface to a C style API. The [python implementation](https://github.com/earth-mover/icechunk/blob/main/icechunk-python/src/lib.rs) does this and should provide a pretty comprehensive guide on what pieces are required.

**Comment by meggart** [comment]
Thanks a lot for pointing to the examples. In oder to start a discussion I shared a few code snippets in #399 @asinghvi17 and others maybe you can have a look?

**Comment by jhamman** [comment]
Hey all — I wanted to share a relevant development here. I've been working on Zarrs.jl, a Julia package that wraps the Rust zarrs library via C FFI, and it already includes Icechunk support.

The package provides:
- Full Zarr V2/V3 read/write support
- Cloud storage backends (S3, GCS, Azure)
- An Icechunk submodule with repository/session management, branching, tagging, and commits
- DiskArrays.jl integration for standard Julia array semantics

For example, here's reading the Dynamical HRRR analysis dataset from an Icechunk repo on S3:

```julia
using Zarrs
using Zarrs.Icechunk

storage = S3Storage(
    bucket    = "dynamical-noaa-hrrr",
    prefix    = "noaa-hrrr-analysis/v0.2.0.icechunk",
    region    = "us-west-2",
    anonymous = true,
)
repo = Repository(storage)
session = readonly_session(repo; branch="main")
g = zopen(session)

# 31 variables, 1799×1059 grid (~3km), 100k+ hourly time steps
temp = g["temperature_2m"]   # (1799, 1059, 100365) Float32
temp[1:10, 1:10, end]        # read a spatial subset at the latest analysis time
```

I wanted to raise this as a possible fast path for Julia users who want to start working with Icechunk. Rather than building and maintaining a standalone C API for Icechunk (which, as discussed in this thread, involves significant complexity around async runtimes, opaque handle lifetimes, etc.), Zarrs.jl bundles Icechunk support into the same shared library that provides the zarrs
bindings. 

The project is still experimental and currently requires a Rust toolchain to build from source (BinaryBuilder support is in progress), but the core functionality — including Icechunk — is working and tested. Would love feedback from folks here, especially @meggart and @asinghvi17 who were looking into the C binding approach.

**Comment by alex-s-gardner** [comment]
> I've been working on Zarrs.jl
Well this is very exciting. Is this in an open repository for poking at?

Found it: https://github.com/earth-mover/Zarrs.jl
